File: leetCode/offer68er-cha-shu-de-zui-jin-gong-gong-zu-xian-lcof/solution.js
class TreeNode {
    constructor(x){
        this.val = x;
        this.left = null;
        this.right = null;
    }
}

function main(){
    let root = new TreeNode(3);
    let node5 = new TreeNode(5);
    let node1 = new TreeNode(1);
    let node6 = new TreeNode(6);
    let node2 = new TreeNode(2);
    let node0 = new TreeNode(0);
    let node8 = new TreeNode(8);
    let node7 = new TreeNode(7);
    let node4 = new TreeNode(4);

    node2.left = node7;
    node2.right = node4;

    node5.left = node6;
    node5.right = node2;

    node1.left = node0;
    node1.right = node8;

    root.left = node5;
    root.right = node1;

    let pPath = [];
    getPath(root, node5, pPath);
    console.log(node5.val,pPath);
    let qPath = [];
    getPath(root, node4, qPath);
    console.log(node4.val,qPath);
}
main();

function getPath(curNode, target, path){
    if(curNode.val == target.val){
        path.push(curNode.val);
        return;
    }
    
    if(curNode.left !=null && path.length == 0) getPath(curNode.left, target, path);
    if(curNode.right != null && path.length == 0) getPath(curNode.right, target, path);

    if(path.length>0) path.push(curNode.val);
}

File: leetCode/offer68er-cha-shu-de-zui-jin-gong-gong-zu-xian-lcof/solution.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
    //     if(root == null || root == p || root == q)return root;

    //     TreeNode left = lowestCommonAncestor(root.left, p, q);
    //     TreeNode right = lowestCommonAncestor(root.right, p, q);
    //     if(left != null && right != null)return root;
    //     if(left != null) return left;
    //     if(right != null) return right;
    //     return null;
    // }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        Stack<TreeNode> pPath = new Stack<>();
        getPath(root, p, pPath);
        Stack<TreeNode> qPath = new Stack<>();
        getPath(root, q, qPath);
        TreeNode ans = root;
        while(!pPath.empty() && !qPath.empty() && pPath.peek().val == qPath.peek().val){
            pPath.pop();
            ans = qPath.pop();
        }
        return ans;
    }

    public void getPath(TreeNode curNode, TreeNode target, Stack<TreeNode> path) {
        if (curNode.val == target.val) {
            path.push(curNode);
            return;
        }

        if (curNode.left != null && path.size()== 0) {
            getPath(curNode.left, target, path);
        }

        if (curNode.right != null && path.size()== 0) {
            getPath(curNode.right, target, path);
        }
        if(path.size()> 0)path.push(curNode);
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }
}

